semtypes/src/main/java/io/ballerina/types/CellAtomicType.java

```diff
@@ -17,34 +17,59 @@
  */
 package io.ballerina.types;
 
+import io.ballerina.types.subtypedata.AllOrNothingSubtype;
+import io.ballerina.types.subtypedata.BddNode;
+
+import static io.ballerina.types.Core.getComplexSubtypeData;
+import static io.ballerina.types.SemTypes.isSubtypeSimple;
+
 /**
  * CellAtomicType node.
  *
+ * @param ty  Type "wrapped" by this cell
+ * @param mut Mutability of the cell
  * @since 2201.10.0
  */
-public final class CellAtomicType implements AtomicType {
-    public final SemType ty;
-    public final CellMutability mut;
-
-    public static final CellAtomicType CELL_ATOMIC_VAL = from(PredefinedType.VAL, CellMutability.CELL_MUT_LIMITED);
-    public static final CellAtomicType CELL_ATOMIC_INNER = from(PredefinedType.INNER, CellMutability.CELL_MUT_LIMITED);
-    public static final CellAtomicType CELL_ATOMIC_NEVER = from(PredefinedType.NEVER, CellMutability.CELL_MUT_LIMITED);
-    public static final CellAtomicType CELL_ATOMIC_INNER_MAPPING = from(
-            Core.union(PredefinedType.MAPPING, PredefinedType.UNDEF), CellMutability.CELL_MUT_LIMITED
-    );
-
-    private CellAtomicType(SemType ty, CellMutability mut) {
-        this.ty = ty;
-        this.mut = mut;
+public record CellAtomicType(SemType ty, CellMutability mut) implements AtomicType {
+
+    public static CellAtomicType from(SemType ty) {
+        if (ty instanceof BasicTypeBitSet bitSet) {
+            return bitSet.bitset == PredefinedType.CELL.bitset ? PredefinedType.CELL_ATOMIC_VAL : null;
+        }
+        assert ty instanceof ComplexSemType;
+        if (!isSubtypeSimple(ty, PredefinedType.CELL)) {
+            return null;
+        }
+        return bddCellAtomicType((Bdd) getComplexSubtypeData((ComplexSemType) ty, BasicTypeCode.BT_CELL),
+                PredefinedType.CELL_ATOMIC_VAL);
+    }
+
+    private static CellAtomicType bddCellAtomicType(Bdd bdd, CellAtomicType top) {
+        if (bdd instanceof AllOrNothingSubtype allOrNothingSubtype) {
+            if (allOrNothingSubtype.isAllSubtype()) {
+                return top;
+            }
+            return null;
+        }
+        BddNode bddNode = (BddNode) bdd;
+        if (bddNode.isSimpleBddNode()) {
+            return cellAtom(bddNode.atom());
+        }
+        return null;
+    }
+
+    private static CellAtomicType cellAtom(Atom atom) {
```

Comment:
In nbal we have this in core as well with default access level.
```suggestion
    static CellAtomicType cellAtomType(Atom atom) {
```

Reply:
Fixed